plugin/session_buffer.py

```diff
@@ -376,5 +392,127 @@ def _present_diagnostics_async(
+            self.session.send_request(request, self._on_semantic_tokens)
+            self.semantic_tokens_pending_request = True
+
+    def _on_semantic_tokens(self, response: Optional[Dict]) -> None:
+        self.semantic_tokens_pending_request = False
+        if response:
+            self.semantic_tokens_result_id = response.get("resultId")
+            self.semantic_tokens_data = response["data"]
+            self._draw_semantic_tokens_async()
+
+    def _on_semantic_tokens_delta(self, response: Optional[Dict]) -> None:
+        self.semantic_tokens_pending_request = False
+        if response:
+            self.semantic_tokens_result_id = response.get("resultId")
+            if "edits" in response:  # response is of type SemanticTokensDelta
+                for semantic_tokens_edit in response["edits"]:
+                    start = semantic_tokens_edit["start"]
+                    end = start + semantic_tokens_edit["deleteCount"]
+                    del self.semantic_tokens_data[start:end]
+                    data = semantic_tokens_edit.get("data")
+                    if data:
+                        self.semantic_tokens_data[start:start] = data
+            elif "data" in response:  # response is of type SemanticTokens
+                self.semantic_tokens_data = response["data"]
+            else:
+                return
+            self._draw_semantic_tokens_async()
+
+    def _draw_semantic_tokens_async(self) -> None:
+        view = self.some_view()
+        if view is None:
+            return
+
+        semantic_tokens_mapping = SEMANTIC_TOKENS_MAP
+        # TODO this mapping better should only be updated when a plugin is registered/unregistered and then stored in the
+        #      Session, and not over and over again for each request
+        if self.session._plugin is not None:
+            semantic_tokens_mapping.update(self.session._plugin.semantic_tokens())
+
+        # TODO compare old and new regions to only remove regions that don't exist anymore
+        for sv in self.session_views:
+            for key in self.semantic_tokens_keys:
+                sv.view.erase_regions(key)
+        self.semantic_tokens.clear()
+        self.semantic_tokens_keys.clear()
+
+        scope_regions = dict()  # type: Dict[str, List[sublime.Region]]
+        prev_line = 0
+        prev_col = 0
+
+        for idx in range(0, len(self.semantic_tokens_data), 5):
+            delta_line = self.semantic_tokens_data[idx]
+            delta_start = self.semantic_tokens_data[idx + 1]
+            length = self.semantic_tokens_data[idx + 2]
+            token_type_encoded = self.semantic_tokens_data[idx + 3]
+            token_modifiers_encoded = self.semantic_tokens_data[idx + 4]
+            line = prev_line + delta_line
+            col = prev_col + delta_start if delta_line == 0 else delta_start
+            a = point_to_offset(Point(line, col), view)
+            b = a + length
```

Comment:
This doesn't feel right. `a` is already in sublime UTF32 space, but this `length` is in LSP UTF16 space I think.